File: Fibonacci Heaps/fibonacci_cascading_cut.c
/* File: fibonacci_cascading_cut.c
   Program: Demonstrate Cascading Cut in Fibonacci Heap
*/

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int key;
    int mark;
    struct Node *parent;
} Node;

Node* createNode(int key)
{
    Node* n = (Node*)malloc(sizeof(Node));
    n->key = key;
    n->mark = 0;
    n->parent = NULL;
    return n;
}

void cascadingCut(Node *y)
{
    Node *z = y->parent;

    if (z != NULL)
    {
        if (y->mark == 0)
        {
            y->mark = 1;
            printf("Node %d marked\n", y->key);
        }
        else
        {
            printf("Cascading cut: node %d cut from parent %d\n", y->key, z->key);
            y->parent = NULL;
            cascadingCut(z);
        }
    }
}

int main()
{
    Node *parent = createNode(10);
    Node *child = createNode(20);

    child->parent = parent;

    cascadingCut(child);
    cascadingCut(child);

    return 0;
}

File: Fibonacci Heaps/fibonacci_cascading_cut.py
# File: fibonacci_cascading_cut.py
# Program: Demonstrate Cascading Cut in Fibonacci Heap

class Node:
    def __init__(self, key):
        self.key = key
        self.mark = False
        self.parent = None


def cascading_cut(y):

    z = y.parent

    if z is not None:

        if not y.mark:
            y.mark = True
            print(f"Node {y.key} marked")

        else:
            print(f"Cascading cut: node {y.key} cut from parent {z.key}")
            y.parent = None
            cascading_cut(z)


parent = Node(10)
child = Node(20)

child.parent = parent

cascading_cut(child)
cascading_cut(child)
